fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -1508,7 +1519,9 @@ public static IndexState writeOnlyIfTooManyRecordsForRebuild(long recordCount, b
                                                     @Nonnull StoreExistenceCheck existenceCheck,
                                                     @Nonnull CompletableFuture<Void> metaDataPreloadFuture) {
         CompletableFuture<Void> subspacePreloadFuture = preloadSubspaceAsync();
-        CompletableFuture<RecordMetaDataProto.DataStoreInfo> storeHeaderFuture = getStoreStateCache().get(this, existenceCheck).thenApply(storeInfo -> {
+        CompletableFuture<FDBRecordStoreStateCacheEntry> cacheEntryFuture = metaDataPreloadFuture.thenCompose(vignore ->
```

Comment:
Do you mean at `upgradeIndexState`? You are right that in most cases, if no upgrade happens, this is just waste of time. Let me know if I understood your request.

Reply:
The more I think about it, the more I agree that the waiting for meta-data must happen at  loading of the states time. Especially that I see there is at least one path in which `preloadRecordStoreStateAsync` is called (in `uncheckedOpenAsync`) before waiting for the meta-data to load at all. 

That said, to me it feels like, the problem is bigger than us waiting for the meta-data too soon, or twice or something. The real problem is how the `FDBRecordStore.Builder` is the one who preloads the meta-data. I propose similar to `preloadSubspaceAsync`, we move `preloadMetaData` to `FDBRecordStore` (or Base). This will ensure we have access to the `future` (if any) at the load time, and if there is no `future` yet, we can start loading the meta-data.  What do you think?

Reply:
@alecgrieser  and I spoke offline on this. I decided to add a `future` to `FDBRecordStore` and its builder to fix this problem.